internal/moduletest/graph/eval_context.go

```diff
@@ -565,43 +568,80 @@ func diagsForEphemeralResources(refs []*addrs.Reference) (diags tfdiags.Diagnost
 	return diags
 }
 
-// UpdateStateFile updates the internal state file for a given state_key value.
-// The TestFileState argument is used to update only the exported fields in the
-// preexisting TestFileState value. Unexported fields' values are preserved
-// from when they are first set while building the test graph.
-//
-// If there isn't a state file for the given state_key then UpdateStateFile will use
-// the TestFileState argument to set its value.
-func (ec *EvalContext) UpdateStateFile(key string, state *TestFileState) {
+func (ec *EvalContext) SetFileState(key string, run *moduletest.Run, state *states.State, reason teststates.StateReason) {
 	ec.stateLock.Lock()
 	defer ec.stateLock.Unlock()
-	oldState, exists := ec.FileStates[key]
-
-	if !exists {
-		ec.FileStates[key] = state
-	}
 
-	ec.FileStates[key] = &TestFileState{
-		File:   state.File,
-		Run:    state.Run,
-		State:  state.State,
-		Reason: state.Reason,
-
-		backend: oldState.backend,
+	current := ec.FileStates[key]
+
+	// Whatever happens we're going to record the latest state for this key.
+	current.State = state
+	current.Manifest.Reason = reason
+
+	if run.Config.SkipCleanup {
+		// if skip cleanup is set on the run block, we're going to track it
+		// as the thing to target regardless of what else might be true.
+		current.Run = run
```

Comment:
checking my thinking here. Because runs of same state are sequential, the run holding this would always be the last one with skip_cleanup=true, right?